apollo-router/src/layers/async_checkpoint.rs

```diff
@@ -67,6 +67,82 @@ where
     }
 }
 
+/// [`Service`] for OneShot (single use) Asynchronous Checkpoints. See [`ServiceBuilderExt::oneshot_checkpoint_async()`](crate::layers::ServiceBuilderExt::oneshot_checkpoint_async()).
+#[allow(clippy::type_complexity)]
+pub struct OneShotAsyncCheckpointService<S, Fut, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request, Error = BoxError> + Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+    Fut: Future<Output = Result<ControlFlow<<S as Service<Request>>::Response, Request>, BoxError>>,
+{
+    inner: Option<S>,
+    checkpoint_fn: Arc<Pin<Box<dyn Fn(Request) -> Fut + Send + Sync + 'static>>>,
+}
+
+impl<S, Fut, Request> OneShotAsyncCheckpointService<S, Fut, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request, Error = BoxError> + Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+    Fut: Future<Output = Result<ControlFlow<<S as Service<Request>>::Response, Request>, BoxError>>,
+{
+    /// Create an `OneShotAsyncCheckpointLayer` from a function that takes a Service Request and returns a `ControlFlow`
+    pub fn new<F>(checkpoint_fn: F, service: S) -> Self
+    where
+        F: Fn(Request) -> Fut + Send + Sync + 'static,
+    {
+        Self {
+            checkpoint_fn: Arc::new(Box::pin(checkpoint_fn)),
+            inner: Some(service),
+        }
+    }
+}
+
+impl<S, Fut, Request> Service<Request> for OneShotAsyncCheckpointService<S, Fut, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request, Error = BoxError> + Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+    Fut: Future<Output = Result<ControlFlow<<S as Service<Request>>::Response, Request>, BoxError>>
+        + Send
+        + 'static,
+{
+    type Response = <S as Service<Request>>::Response;
+
+    type Error = BoxError;
+
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner
+            .as_mut()
+            .expect("One shot must only be called once")
```

Comment:
could this return an error instead of panicking? Panicking means that the entire router is brought down, while an error would only affect the current request. We can still have a `debug_assert!(self.inner.is_some())` though

Reply:
Good suggestion. Bryn also made the same point. I'm going to re-work this.

Reply:
I've re-worked the poll_ready() to improve the resilience: b72a52862d9d80ca82258612ade4f3dad27f8d8e